sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.RequestOptionsProcessor.cs

```diff
@@ -0,0 +1,225 @@
+        private readonly ReadOnlyMemory<PipelinePolicy> _customPerCallPolicies;
+
+        // Custom per-try policies used for the scope of the method invocation.
+        private readonly ReadOnlyMemory<PipelinePolicy> _customPerTryPolicies;
+
+        // Custom per-try policies used for the scope of the method invocation.
+        private readonly ReadOnlyMemory<PipelinePolicy> _customBeforeTransportPolicies;
+
+        private PolicyEnumerator? _enumerator;
+
+        public RequestOptionsProcessor(
+            ReadOnlyMemory<PipelinePolicy> fixedPolicies,
+            ReadOnlyMemory<PipelinePolicy> perCallPolicies,
+            ReadOnlyMemory<PipelinePolicy> perTryPolicies,
+            ReadOnlyMemory<PipelinePolicy> beforeTransportPolicies,
+            int perCallIndex,
+            int perTryIndex,
+            int beforeTransportIndex)
+        {
+            Debug.Assert(perCallIndex <= fixedPolicies.Length);
+            Debug.Assert(perTryIndex <= fixedPolicies.Length);
+            Debug.Assert(beforeTransportIndex <= fixedPolicies.Length);
+            Debug.Assert(perCallIndex <= perTryIndex);
+            Debug.Assert(perTryIndex <= beforeTransportIndex);
+
+            _fixedPolicies = fixedPolicies;
+            _customPerCallPolicies = perCallPolicies;
+            _customPerTryPolicies = perTryPolicies;
+            _customBeforeTransportPolicies = beforeTransportPolicies;
+
+            _perCallIndex = perCallIndex;
+            _perTryIndex = perTryIndex;
+            _beforeTransportIndex = beforeTransportIndex;
+
+            _length = _fixedPolicies.Length +
+                _customPerCallPolicies.Length +
+                _customPerTryPolicies.Length +
+                _customBeforeTransportPolicies.Length;
+        }
+
+        public PipelinePolicy this[int index]
+        {
+            get
+            {
+                TryGetPolicy(index, out PipelinePolicy policy);
+                return policy;
+            }
+        }
+
+        public int Count => _length;
+
+        public IEnumerator<PipelinePolicy> GetEnumerator()
+            => _enumerator ??= new(this);
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => GetEnumerator();
+
+        /// <summary>
+        /// This custom pipeline is divided into seven segments by the per-call,
+        /// per-try, and before-transport indexes:
+        ///
+        /// [FixedPerCall] [CustomPerCall][FixedPerTry] [CustomPerTry][FixedPerTransport] [CustomBeforeTransport][Transport]
+        ///               ^_perCallIndex               ^_perTryIndex                     ^_beforeTransport
+        ///
+        /// This method returns the next policy in the customized pipeline
+        /// sequence and maintains state by incrementing the _current counter
+        /// after each "next policy" is returned.
+        /// </summary>
+        private bool TryGetPolicy(int index, out PipelinePolicy policy)
```

Comment:
nit: would it make sense to pre-calculate these boundaries once on construction, rather than on each index call?

Reply:
Would you be willing to share an example of what you have in mind?

Reply:
Initially I was thinking we could pre-calculate all the boundary offsets along with the total policies prior to those boundaries, but even that is more complicated than creating a superset `ReadOnlyMemory<PipelinePolicy>`containing the whole chain that is agnostic to the segment types. I'm not even sure if we need to keep them in separate collections to support this implementation, since we have the offsets if we ever need to know which one they belong to. If we did want to keep the individual segment collections for some reason, I think we'd still be making a good perf tradeoff between additional memory and runtime perf.

Reply:
I actually spent a lot of time looking at how we could precompute compressed-storage lookup tables to avoid these TryGetMethods and also the processor branching instruction costs they incur.  I made some progress and if time allows I may try to apply some of what I developed to see if we can apply those perf gains to this.  I also have an outstanding issue tracking doing a perf comparison across the Azure.Core "copy into a new policy ROM" approach vs. this one ... so we may come back to this in the next sprint if time allows.

Reply:
Tracking the perf investigation in this issue: https://github.com/Azure/azure-sdk-for-net/issues/40754